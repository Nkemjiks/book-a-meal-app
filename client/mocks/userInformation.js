const userInformation = {
  user: {
    id: '4094ad8a-be5e-43a4-8c58-3e0c680f5fc9',
    fullName: 'Andela Bayo',
    email: 'mbonunkemjika@ymail.com',
    phoneNumber: '2334',
    role: 'caterer',
    address: '12B agege',
  },
  error: null,
};

export default userInformation;
