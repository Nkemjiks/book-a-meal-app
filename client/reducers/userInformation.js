import {
  SIGNUP_USER_SUCCESS,
  SIGNUP_USER_FAILURE,
  LOGIN_USER_FAILURE,
  LOGIN_USER_SUCCESS,
  GET_USER_DETAILS_SUCCESS,
  USER_ROLE_UPDATE_SUCCESS,
  USER_ROLE_UPDATE_FAILURE,
  LOGOUT_USER_SUCCESS,
} from '../actionTypes';

const initialState = {
  user: {},
  error: null,
};

/**
 * Reducer that handles user related requests
 *(signup, login, logout, user role update)
 *
 * @param {Object} state initial state for the userInformation section of the store
 * @param {Object} action the dispatched action
 *
 * @returns {Object} new state of the userInformation section of the store
 */
export default (state = initialState, action) => {
  switch (action.type) {
    case SIGNUP_USER_SUCCESS:
      return {
        ...state,
        user: action.payload,
        error: null,
      };
    case SIGNUP_USER_FAILURE:
      return {
        ...state,
        user: {},
        error: action.payload,
      };
    case LOGIN_USER_SUCCESS:
      return {
        ...state,
        user: action.payload,
        error: null,
      };
    case LOGIN_USER_FAILURE:
      return {
        ...state,
        user: {},
        error: action.payload,
      };
    case GET_USER_DETAILS_SUCCESS:
      return {
        ...state,
        user: action.payload,
        error: null,
      };
    case USER_ROLE_UPDATE_SUCCESS:
      return {
        ...state,
        user: action.payload,
        error: null,
      };
    case USER_ROLE_UPDATE_FAILURE:
      return {
        ...state,
        user: {},
        error: action.payload,
      };
    case LOGOUT_USER_SUCCESS:
      return {
        ...state,
        user: {},
        error: null,
      };
    default:
      return state;
  }
};
