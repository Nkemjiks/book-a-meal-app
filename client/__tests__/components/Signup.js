import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme, { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import ConnectedSignup, { Signup } from '../../components/Signup';
import userInformation from '../../mocks/userInformation';

const props = {
  signupAction: jest.fn(),
  history: {
    push: jest.fn(),
  },
};


Enzyme.configure({ adapter: new Adapter() });

const mockStore = configureMockStore();
const store = mockStore({ userInformation });
const wrapper = shallow(<Signup {...props} />);


describe('Signup Component', () => {
  it('should render unconnected component properly', () => {
    expect(wrapper.find('#signup-component').length).toBe(1);
    expect(wrapper.find('#signup-container').length).toBe(1);
    expect(wrapper.find('h1').first().text()).toBe('Sign Up Now');
    expect(wrapper.find('form').length).toBe(1);
    expect(wrapper.find('input').length).toBe(5);
    expect(wrapper.find('button').length).toBe(1);
  });
  it('should call the handle change method to update state', () => {
    const handleChangeSpy = jest.spyOn(wrapper.instance(), 'handleChange');
    const firstEvent = {
      target: {
        name: 'fullName',
        value: 'Mbonu',
      },
    };
    const secondEvent = {
      target: {
        name: 'phoneNumber',
        value: 90678099,
      },
    };
    wrapper.instance().handleChange(firstEvent);
    wrapper.instance().handleChange(secondEvent);
    expect(wrapper.state().fullName).toBe('Mbonu');
    expect(wrapper.state().phoneNumber).toBe(90678099);
    expect(handleChangeSpy).toHaveBeenCalled();
  });
  it('should not call the action with an empty email or password', () => {
    wrapper.setState({
      fullName: 'Jeremy John',
      email: '',
      phoneNumber: '',
      password: 'tester',
      address: '12B, Agege Road',
    });
    const handleSubmitSpy = jest.spyOn(wrapper.instance(), 'handleSubmit');
    const signinSpy = jest.spyOn(wrapper.instance().props, 'signupAction');
    const submitButton = wrapper.find('.button').first();
    const event = {
      preventDefault: jest.fn(),
    };
    submitButton.simulate('click', event);
    wrapper.instance().handleSubmit(event);
    expect(handleSubmitSpy).toHaveBeenCalled();
    expect(signinSpy).not.toHaveBeenCalled();
  });

  it('should call the action with a valid email and password', () => {
    wrapper.setState({
      fullName: 'Jeremy John',
      email: 'jeremy@gmail.com',
      phoneNumber: 9072873939,
      password: 'tester',
      address: '12B, Agege Road',
    });
    const handleSubmitSpy = jest.spyOn(wrapper.instance(), 'handleSubmit');
    const signinSpy = jest.spyOn(wrapper.instance().props, 'signupAction');
    const submitButton = wrapper.find('.button').first();
    const event = {
      preventDefault: jest.fn(),
    };
    submitButton.simulate('click', event);
    wrapper.instance().handleSubmit(event);
    expect(handleSubmitSpy).toHaveBeenCalled();
    expect(signinSpy).toHaveBeenCalled();
  });

  it('should render connected component properly', () => {
    const connectedWrapper = shallow(<ConnectedSignup {...props} store={store} />);
    expect(connectedWrapper.length).toBe(1);
  });
});
