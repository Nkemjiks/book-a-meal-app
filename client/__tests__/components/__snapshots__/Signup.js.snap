// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup Component renders correctly 1`] = `
<div
  id="signup-component"
>
  <div
    id="signup-container"
  >
    <h1>
      Sign Up Now
    </h1>
    <form>
      <input
        name="fullName"
        onChange={[Function]}
        placeholder="Full Name"
        type="text"
        value=""
      />
      <input
        name="email"
        onChange={[Function]}
        placeholder="Email Address"
        type="email"
        value=""
      />
      <input
        name="phoneNumber"
        onChange={[Function]}
        placeholder="Phone Number"
        type="tel"
        value=""
      />
      <input
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
      <input
        name="address"
        onChange={[Function]}
        placeholder="Address"
        type="text"
        value=""
      />
      <button
        className="button"
        onClick={[Function]}
      >
        SIGN UP
      </button>
    </form>
  </div>
</div>
`;
